File: workbook/ch07/data/bintree.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        if not self.root:
            self.root = Node(key)
            return
        queue = [self.root]
        while queue:
            temp = queue.pop(0)
            if not temp.left:
                temp.left = Node(key)
                return
            else:
                queue.append(temp.left)
            if not temp.right:
                temp.right = Node(key)
                return
            else:
                queue.append(temp.right)

    def search(self, root, key):
        if not root:
            return False
        if root.key == key:
            return True
        return self.search(root.left, key) or self.search(root.right, key)

    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.key, end=" ")
            self.inorder(root.right)

    def preorder(self, root):
        if root:
            print(root.key, end=" ")
            self.preorder(root.left)
            self.preorder(root.right)

    def postorder(self, root):
        if root:
            self.postorder(root.left)
            self.postorder(root.right)
            print(root.key, end=" ")

    def level_order(self):
        if not self.root:
            return
        queue = [self.root]
        while queue:
            temp = queue.pop(0)
            print(temp.key, end=" ")
            if temp.left:
                queue.append(temp.left)
            if temp.right:
                queue.append(temp.right)

    def height(self, root):
        if not root:
            return 0
        return 1 + max(self.height(root.left), self.height(root.right))

    def delete(self, key):
        if not self.root:
            return None
        if self.root.key == key and not self.root.left and not self.root.right:
            self.root = None
            return

        queue = [self.root]
        target, last = None, None

        while queue:
            last = queue.pop(0)
            if last.key == key:
                target = last
            if last.left:
                queue.append(last.left)
            if last.right:
                queue.append(last.right)

        if target:
            target.key = last.key
            queue = [self.root]
            while queue:
                temp = queue.pop(0)
                if temp.left == last:
                    temp.left = None
                    return
                elif temp.left:
                    queue.append(temp.left)
                if temp.right == last:
                    temp.right = None
                    return
                elif temp.right:
                    queue.append(temp.right)

# example
tree = BinaryTree()
for key in [10, 20, 30, 40, 50, 60, 70]:
    tree.insert(key)

print("Inorder Traversal:")
tree.inorder(tree.root)
print("\nPreorder Traversal:")
tree.preorder(tree.root)
print("\nPostorder Traversal:")
tree.postorder(tree.root)
print("\nLevel-order Traversal:")
tree.level_order()
print("\nHeight of Tree:", tree.height(tree.root))

tree.delete(30)
print("\nInorder Traversal after deleting 30:")
tree.inorder(tree.root)

File: workbook/ch07/data/common/bintree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int key;
    struct Node *left, *right;
} Node;

Node *createNode(int key) {
    Node *node = (Node *)malloc(sizeof(Node));
    node->key = key;
    node->left = node->right = NULL;
    return node;
}

Node *insert(Node *root, int key) {
    if (!root)
        return createNode(key);
    if (key < root->key)
        root->left = insert(root->left, key);
    else
        root->right = insert(root->right, key);
    return root;
}

Node *search(Node *root, int key) {
    if (!root || root->key == key)
        return root;
    if (key < root->key)
        return search(root->left, key);
    return search(root->right, key);
}

void inorder(Node *root) {
    if (root) {
        inorder(root->left);
        printf("%d ", root->key);
        inorder(root->right);
    }
}

void preorder(Node *root) {
    if (root) {
        printf("%d ", root->key);
        preorder(root->left);
        preorder(root->right);
    }
}

void postorder(Node *root) {
    if (root) {
        postorder(root->left);
        postorder(root->right);
        printf("%d ", root->key);
    }
}

int height(Node *root) {
    if (!root)
        return 0;
    int leftHeight = height(root->left);
    int rightHeight = height(root->right);
    return (leftHeight > rightHeight ? leftHeight : rightHeight) + 1;
}

Node *minValueNode(Node *node) {
    Node *current = node;
    while (current && current->left)
        current = current->left;
    return current;
}

Node *deleteNode(Node *root, int key) {
    if (!root)
        return root;
    if (key < root->key)
        root->left = deleteNode(root->left, key);
    else if (key > root->key)
        root->right = deleteNode(root->right, key);
    else {
        if (!root->left) {
            Node *temp = root->right;
            free(root);
            return temp;
        } else if (!root->right) {
            Node *temp = root->left;
            free(root);
            return temp;
        }
        Node *temp = minValueNode(root->right);
        root->key = temp->key;
        root->right = deleteNode(root->right, temp->key);
    }
    return root;
}

void levelOrder(Node *root) {
    if (!root) return;
    Node *queue[100];
    int front = 0, rear = 0;
    queue[rear++] = root;
    while (front < rear) {
        Node *temp = queue[front++];
        printf("%d ", temp->key);
        if (temp->left)
            queue[rear++] = temp->left;
        if (temp->right)
            queue[rear++] = temp->right;
    }
}

int main() {
    Node *root = NULL;
    root = insert(root, 10);
    insert(root, 20);
    insert(root, 5);
    insert(root, 15);
    insert(root, 30);

    printf("Inorder Traversal:\n");
    inorder(root);
    printf("\nPreorder Traversal:\n");
    preorder(root);
    printf("\nPostorder Traversal:\n");
    postorder(root);
    printf("\nLevel-order Traversal:\n");
    levelOrder(root);
    printf("\nHeight of Tree: %d\n", height(root));

    root = deleteNode(root, 15);
    printf("\nInorder after deleting 15:\n");
    inorder(root);
    printf("\n");

    return 0;
}